### PR TITLE
Major restructuring of autonomy logic.

### PR DESCRIPTION
**Pull request by angustomlinson**
- Used the state pattern to reorganize autonomy.
- Camera to world coordinates conversion complete.
- Driving logic complete.
- State machine complete.

### CODE REVIEW

File: Autonomy/app/src/main/java/nasa_rmc/autonomy/Data.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
+package nasa_rmc.autonomy;
+
+import com.google.atap.tangoservice.TangoPoseData;
+
+/**
+ * Created by atomlinson on 3/31/17.
+ */
+
+public class Data {
```
Comment (JWhitney406) [inline Autonomy/app/src/main/java/nasa_rmc/autonomy/Data.java]: Oh yes.... +10000

File: Autonomy/app/src/main/java/nasa_rmc/autonomy/ForwardingPrefix.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package nasa_rmc.autonomy;
+
+/**
+ * Created by atomlinson on 4/3/17.
+ */
+
+public enum ForwardingPrefix {
```
Comment (JWhitney406) [inline Autonomy/app/src/main/java/nasa_rmc/autonomy/ForwardingPrefix.java]: Maybe add comments on how the Pi handles each of these, or what they mean.

File: Autonomy/app/src/main/java/nasa_rmc/autonomy/DriveState.java [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+            if (hasArrivedAtCoordinates(itinerary.getNextCoordinates())) {
+                status = "At final destination.";
+                itinerary.arrivedAtCoordinates();
+            }
+
+            TimeUnit.MILLISECONDS.sleep(DRIVE_COMMAND_DELAY);
+        }
+
+        logicContext.setLogicState(logicContext.getDigState());
```
Comment (JWhitney406) [inline Autonomy/app/src/main/java/nasa_rmc/autonomy/DriveState.java]: Don't we need logic here to decide whether we transition to dig state, or dump state? I don't see how we get to the dump state currently.

File: Autonomy/app/src/main/java/nasa_rmc/autonomy/ForwardingPrefix.java (hunk 2 above)
Comment (angustomlinson) [inline Autonomy/app/src/main/java/nasa_rmc/autonomy/ForwardingPrefix.java]: Will do. I'm going to go through and add comments to everything once I get the structure completely done.

File: Autonomy/app/src/main/java/nasa_rmc/autonomy/DriveState.java (hunk 3 above)
Comment (angustomlinson) [inline Autonomy/app/src/main/java/nasa_rmc/autonomy/DriveState.java]: @JWhitney406 Ah, very good point. I hadn't gotten to that point in the state cycle yet, so it had never crossed my mind.

File: Autonomy/app/src/main/java/nasa_rmc/autonomy/AutonomyActivity.java [hunk 4]
```diff
@@ -248,21 +232,16 @@ public void onPoseAvailable(TangoPoseData pose) {
                 runOnUiThread(new Runnable() {
```
Comment (JWhitney406) [inline Autonomy/app/src/main/java/nasa_rmc/autonomy/AutonomyActivity.java]: Incredible... we have a UI thread that actually does UI things.

File: Autonomy/app/src/main/java/nasa_rmc/autonomy/logic/logicState/DriveState.java [hunk 5]
```diff
@@ -59,8 +71,15 @@ public void run() throws InterruptedException {
             TimeUnit.MILLISECONDS.sleep(DRIVE_COMMAND_DELAY);
         }
 
-        logicContext.setLogicState(logicContext.getDigState());
-        logicContext.getLogicState().run();
+        if (itinerary.getItineraryPurpose() == Itinerary.ItineraryPurpose.DIG) {
```
Comment (JWhitney406) [inline Autonomy/app/src/main/java/nasa_rmc/autonomy/logic/logicState/DriveState.java]: Sweet. +1

**Pull request merged**